### PR TITLE
docs: vault ca provider patch upgrade guidance

### PR DESCRIPTION
**Pull request by jkirschner-hashicorp**
Status: looking for
- [x] engineering to review for accuracy
- [x] education to review for clarity

Follow on work:
- Add a separate PR to note this is needed when upgrading to 1.14 (if not already handled)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,10 @@ BUG FIXES:
 
 ## 1.13.2 (September 20, 2022)
 
+BREAKING CHANGES:
+
+* ca: If using Vault as the service mesh CA provider, and the Vault PKI is managed by Vault itself rather than Consul, the Vault policy used by Consul now requires the `update` capability on the intermediate PKI's tune mount configuration endpoint, such as `/sys/mounts/connect_inter/tune`. This change was necessary to address [[GH-14516](https://github.com/hashicorp/consul/issues/14516)]. Refer to [upgrade guidance](https://www.consul.io/docs/upgrading/upgrade-specific#modify-vault-policy-for-vault-ca-provider) for more information.
```
Comment (jkirschner-hashicorp) [inline CHANGELOG.md:11]: Note to self: verify these links work

File: website/content/docs/connect/ca/vault.mdx [hunk 2]
```diff
@@ -227,6 +227,11 @@ path "/sys/mounts/connect_inter" {
   capabilities = [ "read" ]
 }
 
+# Needed for Consul 1.11+
```
Comment (jkirschner-hashicorp) [inline website/content/docs/connect/ca/vault.mdx:230]: I intentionally simplified this to 1.11+.

Saying 1.14.0+, 1.13.2+, 1.12.5+, and 1.11.9+ is more "technically accurate", but also more confusing and hard to follow. I think it's safest to just say "1.11+".

File: website/content/docs/upgrading/upgrade-specific.mdx [hunk 3]
```diff
@@ -36,6 +36,25 @@ For most environments, the Envoy communication to Consul is loop-back only and d
...
 
+### Modify Vault Policy for Vault CA Provider
+
+If using the Vault CA provider,
+you must modify the Vault policy used by Consul to interact with Vault
+if Vault is managing the PKI paths itself.
+The policy must include the `update` capability on the intermediate PKI's tune mount configuration endpoint
+at path `/sys/mounts/<intermediate_pki_mount_name>/tune`.
+For an updated example Vault policy for use with Vault-managed PKI paths,
+refer to the [Vault CA provider documentation](/docs/connect/ca/vault#vault-managed-pki-paths).
```
Comment (jkirschner-hashicorp) [inline website/content/docs/upgrading/upgrade-specific.mdx:47]: Note to self: verify the new links on this page work

File: website/content/docs/upgrading/upgrade-specific.mdx [hunk 4]
```diff
@@ -178,7 +197,19 @@ The [`disable_compat_19`](/docs/agent/options#telemetry-disable_compat_1.9) tele
 In prior Consul versions (1.10.x through 1.11.x), the config defaulted to `false`. In 1.12.x it defaulted to `true`.
 If you were using this flag, you must remove it before upgrading.
 
-## Consul 1.12.0
+### Modify Vault Policy for Vault CA Provider
+
+If upgrading to Consul 1.13.2 or later,
```
Comment (jkirschner-hashicorp) [inline website/content/docs/upgrading/upgrade-specific.mdx:202]: Note to reviewers: I wasn't sure how to best copy the same guidance into multiple places that it's needed. Currently, this PR is just referencing the 1.14 content from the 1.11-1.13 sections (to avoid duplicate content). I'll obviously need to modify that somewhat to merge into 1.13 immediately (not waiting for 1.14 docs release to get this important info out).

**Comment by thomashashi** [comment]
I do not believe this is limited to Vault-managed PKI paths: I have a test setup where the paths are Consul managed, and while I was restarting it this morning to look at something else I got this error message:

```
Oct 06 14:08:30 consul-server-0-dc1 consul[12920]: 2022-10-06T14:08:30.983Z [INFO]  connect.ca.vault: Successfully renewed token for Vault provider
Oct 06 14:08:30 consul-server-0-dc1 consul[12920]: 2022-10-06T14:08:30.983Z [ERROR] connect.ca: Failed to initialize Connect CA: routine="CA initialization" error="error configuring provider: Error making API request.
Oct 06 14:08:30 consul-server-0-dc1 consul[12920]: URL: POST http://127.0.0.1:8200/v1/sys/mounts/connect_dc1_inter/tune
Oct 06 14:08:30 consul-server-0-dc1 consul[12920]: Code: 403. Errors:
Oct 06 14:08:30 consul-server-0-dc1 consul[12920]: * 1 error occurred:
Oct 06 14:08:30 consul-server-0-dc1 consul[12920]:         * permission denied
Oct 06 14:08:30 consul-server-0-dc1 consul[12920]: "
```

I had to add the following section to the `consul-managed-pki-policy.hcl` from https://www.consul.io/docs/connect/ca/vault#consul-managed-pki-paths

```
path "/sys/mounts/connect_dc1_inter/tune" {
  capabilities = [ "update" ]
}
```

(in this case, `connect_dc1_inter` is what I'm using in place of `connect_inter`).

**Comment by thomashashi** [comment]
FYI on my last comment:
Consul v1.13.2
Vault v1.11.4

**Comment by thomashashi** [comment]
To add more context, when I restart my test Consul cluster, it wants to update the intermediate path. If it can't because of this permissions issue, then the CA provider isn't running and Consul cannot generate or do anything with certificates. For example, I started my Consul cluster with a Vault token without the `update` permission on `/sys/mounts/connect_dc1_inter/tune`, it reports the error:

```
Oct 06 15:14:44 consul-server-0-dc1 consul[13531]: 2022-10-06T15:14:44.193Z [ERROR] connect.ca: Failed to initialize Connect CA: routine="CA initialization" error="error configuring provider: Error making API request.
Oct 06 15:14:44 consul-server-0-dc1 consul[13531]: URL: POST http://127.0.0.1:8200/v1/sys/mounts/connect_dc1_inter/tune
Oct 06 15:14:44 consul-server-0-dc1 consul[13531]: Code: 403. Errors:
Oct 06 15:14:44 consul-server-0-dc1 consul[13531]: * 1 error occurred:
Oct 06 15:14:44 consul-server-0-dc1 consul[13531]:         * permission denied
```

and if a client comes along and tries to get a leaf certificate:

```
$ curl -v http://127.0.0.1:8500/v1/agent/connect/ca/leaf/alsdkjfaldkjfads
* Expire in 0 ms for 6 (transfer 0x5635adb6c0f0)
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Expire in 200 ms for 4 (transfer 0x5635adb6c0f0)
* Connected to 127.0.0.1 (127.0.0.1) port 8500 (#0)
> GET /v1/agent/connect/ca/leaf/alsdkjfaldkjfads HTTP/1.1
> Host: 127.0.0.1:8500
> User-Agent: curl/7.64.0
> Accept: */*
> 
< HTTP/1.1 500 Internal Server Error
< Vary: Accept-Encoding
< X-Consul-Default-Acl-Policy: allow
< Date: Thu, 06 Oct 2022 15:15:34 GMT
< Content-Length: 76
< Content-Type: text/plain; charset=utf-8
< 
* Connection #0 to host 127.0.0.1 left intact
CA is uninitialized and unable to sign certificates yet: no root certificate
```

which shows up on the server side as:

```
Oct 06 15:15:34 consul-server-0-dc1 consul[13531]: 2022-10-06T15:15:34.676Z [ERROR] agent.http: Request error: method=GET url=/v1/agent/connect/ca/leaf/alsdkjfaldkjfads from=127.0.0.1:45302 error="CA is uninitialized and unable to sign certificates yet: no root certificate"
```

As soon as I update the policy on the Vault side I see this:

```
Oct 06 15:16:34 consul-server-0-dc1 consul[13531]: 2022-10-06T15:16:34.786Z [INFO]  connect.ca.vault: Successfully renewed token for Vault provider
Oct 06 15:16:34 consul-server-0-dc1 consul[13531]: 2022-10-06T15:16:34.813Z [INFO]  connect.ca: CA provider config updated
Oct 06 15:18:28 consul-server-0-dc1 consul[13531]: 2022-10-06T15:18:28.204Z [INFO]  connect.ca: Successfully initialized the Connect CA
Oct 06 15:18:28 consul-server-0-dc1 consul[13531]: 2022-10-06T15:18:28.204Z [INFO]  agent.leader: started routine: routine="intermediate cert renew watch"
Oct 06 15:18:28 consul-server-0-dc1 consul[13531]: 2022-10-06T15:18:28.204Z [INFO]  agent.leader: stopped routine: routine="CA initialization"
```

and now I can generate leaf certificates.

File: CHANGELOG.md [hunk 5]
```diff
@@ -6,6 +6,10 @@ BUG FIXES:
 
 ## 1.13.2 (September 20, 2022)
 
+BREAKING CHANGES:
+
+* ca: If using Vault as the service mesh CA provider, the Vault policy used by Consul now requires the `update` capability on the intermediate PKI's tune mount configuration endpoint, such as `/sys/mounts/connect_inter/tune`. This change was necessary to address [[GH-14516](https://github.com/hashicorp/consul/issues/14516)]. Refer to [upgrade guidance](https://www.consul.io/docs/upgrading/upgrade-specific#modify-vault-policy-for-vault-ca-provider) for more information.
```
Comment (jkirschner-hashicorp) [inline CHANGELOG.md:11]: Note to self: replace all these links with https://www.consul.io/docs/connect/ca/vault#vault-acl-policies

File: website/content/docs/upgrading/upgrade-specific.mdx [hunk 6]
```diff
@@ -170,6 +171,29 @@ such as with flags or environment variables like
...
+If using the Vault CA provider,
+you must modify the Vault policy used by Consul to interact with Vault
+so that certificates required for service mesh operation can still be generated.
+The policy must include the `update` capability on the intermediate PKI's tune mount configuration endpoint
+at path `/sys/mounts/<intermediate_pki_mount_name>/tune`.
+Refer to the [Vault CA provider documentation](/docs/connect/ca/vault#vault-managed-pki-paths)
+for updated example Vault policies for use with Vault-managed or Consul-managed PKI paths.
+
+You are using the Vault CA provider if either of the following configurations exists:
+- The Consul agent configuration option [`connect.ca_provider`](/docs/agent/config/config-files#connect_ca_provider) is set to `vault`, or
```
Comment (jkirschner-hashicorp) [inline website/content/docs/upgrading/upgrade-specific.mdx:185]:
```suggestion
- The Consul server agent configuration option [`connect.ca_provider`](/docs/agent/config/config-files#connect_ca_provider) is set to `vault`, or
```

**Review by kyhavlov — APPROVED** [review]

**Comment by jkirschner-hashicorp** [comment]
Thanks @boruszak ! I'll ensure I resolve those "notes to self" before merging. Wanted to see if there were any other review comments from eng / edu first, but it looks like there aren't any this time around!

File: CHANGELOG.md (hunk 5 above)
Comment (jkirschner-hashicorp) [inline CHANGELOG.md:11]: Note to self: document that we're looking into making this better in an upcoming patch

Comment (jkirschner-hashicorp) [reply CHANGELOG.md:11]: I put this comment in the wrong place (was meant for the upgrade guidance), but I've resolved it nonetheless

File: website/content/docs/upgrading/upgrade-specific.mdx [hunk 7]
```diff
@@ -170,6 +171,29 @@ such as with flags or environment variables like
...
+- The Consul on Kubernetes Helm Chart [`global.secretsBackend.vault.connectCA`](/docs/k8s/helm#v-global-secretsbackend-vault-connectca) value is configured.
+
+Though this guidance is listed in the 1.13.x section, it applies to all of the following release series:
+- Consul 1.13.x: applies to 1.13.2+
+- Consul 1.12.x: applies to 1.12.5+
+- Consul 1.11.x: applies to 1.11.9+
+
+We recommend modifying the Vault policy before upgrading to Consul 1.11 or later
+to ensure your organization does accidentally miss this guidance when performing subsequent upgrades,
```
Comment (jkirschner-hashicorp) [inline website/content/docs/upgrading/upgrade-specific.mdx:194]:
```suggestion
to ensure your organization does not accidentally miss this guidance when performing subsequent upgrades,
```

**Comment by jkirschner-hashicorp** [comment]
I need to manually backport to 1.13 and stable-website

**Comment by jkirschner-hashicorp** [comment]
Backport to `stable-website`: https://github.com/hashicorp/consul/pull/14914

**Comment by jkirschner-hashicorp** [comment]
Backport to `release/1.13.x`: https://github.com/hashicorp/consul/pull/14915
